### ISSUE TITLE
PaperDb crashing with Kotlin by lazy

### ISSUE THREAD

**Issue by ghosalmartin** [pilgr/Paper#109]
When attempting to serialize a data class with Kotlin by lazy this occurs 

```io.reactivex.exceptions.UndeliverableException: java.lang.StackOverflowError: stack size 1038KB
    at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:349)
    at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:69)
    at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
    at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.StackOverflowError: stack size 1038KB
    at java.lang.ThreadLocal.get(ThreadLocal.java:161)
    at dalvik.system.BlockGuard.getThreadPolicy(BlockGuard.java:145)
    at libcore.io.BlockGuardOs.write(BlockGuardOs.java:344)
    at libcore.io.IoBridge.write(IoBridge.java:553)
    at java.io.FileOutputStream.write(FileOutputStream.java:326)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:186)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)
    at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:251)
    at com.esotericsoftware.kryo.io.Output.write(Output.java:219)
    at com.esotericsoftware.kryo.io.Output.flush(Output.java:185)
    at com.esotericsoftware.kryo.io.OutputChunked.flush(OutputChunked.java:59)
    at com.esotericsoftware.kryo.io.Output.require(Output.java:164)```

A test project was made here replicating the issue 

https://github.com/ghosalmartin/paperdbtest

Using the latest paperDB

**Comment by dandroid6000** [comment]
TL;DR — Try annotating the property with `@delegate:Transient` in order to skip serializing the lazy field.  Also, if your class does not supply default values for all parameters, you must supply a parameterless secondary constructor (this can be declared private) for deserialization to work.


I also ran into this and did some digging into the bytecode (Tools -> Kotlin -> Show Kotlin Bytecode).  Here's what I found:

Decompiling the data model bytecode shows this for your sample data model:
```Java
// $FF: synthetic field
   static final KProperty[] $$delegatedProperties = new KProperty[]{(KProperty)Reflection.property1(new PropertyReference1Impl(Reflection.getOrCreateKotlinClass(Model.class), "foo", "getFoo()Ljava/lang/String;"))};
   private final Lazy foo$delegate;

   private final String getFoo() {
      Lazy var1 = this.foo$delegate;
      KProperty var3 = $$delegatedProperties[0];
      return (String)var1.getValue();
   }

   ...

   public Model(@NotNull String test) {
      Intrinsics.checkParameterIsNotNull(test, "test");
      super();
      this.test = test;
      this.foo$delegate = LazyKt.lazy((Function0)(new Function0() {
         // $FF: synthetic method
         // $FF: bridge method
         public Object invoke() {
            return this.invoke();
         }

         @NotNull
         public final String invoke() {
            return Model.this.getTest();
         }
      }));
   }

   // $FF: synthetic method
   public Model(String var1, int var2, DefaultConstructorMarker var3) {
      if((var2 & 1) != 0) {
         var1 = "";
      }

      this(var1);
   }

   public Model() {
      this((String)null, 1, (DefaultConstructorMarker)null);
   }
...
```

Essentially that lazy property gets turned in to a Lazy instance and a getter, and the Lazy instance is instantiated in the single-parameter constructor.

PaperDb uses Kryo for serialization/deserialization, and Kryo doesn't support de/serializing a Lazy object.  This is what causes the crash you're seeing.

But in your case, you don't need to save the value of `foo`, you just need to regenerate it by letting the lazy code do its thing.  Fortunately you can skip de/serialization by annotating the lazy property with `@delegate:Transient`.  This solves your issue.

However!  This only one piece of the puzzle.  Kryo [uses reflection to instantiate your object](https://github.com/EsotericSoftware/kryo#object-creation) by calling its **parameterless constructor**.  Your case works because you are supplying a default value for your parameter.  This generates a parameterless constructor in your bytecode.  You'll see above that the parameterless constructor calls the single-parameter constructor which instantiates the Lazy object.

In my case, I did not have default parameters, and so no parameterless constructor was generated.  As a result, when Kryo deserialized my object, the Lazy object was never instantiated, and so when the property is accessed, its getter is called, and `return (String)var1.getValue();` throws an NPE because var1 (the Lazy object) is null.

Kryo actually recommends supplying a private parameterless constructor as a means of instantiating your object.  You can set property values to anything, since they'll be overwritten by deserialization.

In summary, a couple extra lines of code can fix the problem, though it does seem like something that Paper should have in their docs.

**Comment by ghosalmartin** [comment]
Ah very thoroughly and worthwhile answer :) Thanks for taking the time to explain the issue. `@delegate:Transient` does indeed work, maybe if its added to docs somewhere for others to avoid the same pitfall since I'd be closing the issue

**Comment by pilgr** [comment]
Hey guys, it's already in the doc
https://github.com/pilgr/Paper#exclude-fields

**Comment by dandroid6000** [comment]
Yeah, the transient keyword is.  The Kotlin delegate version isn't, but that's just a matter of knowing how to apply annotations to delegate properties.

I was more talking about the need for a parameterless constructor in your model object when using Kotlin delegates, due to the fact that the delegate object is instantiated in the primary constructor and will be broken if it's not called.  I had to decompile bytecode and dig into the Kryo documentation to figure this out, which is more than some may be willing to do get their model objects to work with this library.

**Comment by pilgr** [comment]
Yeah, many thanks @dandroid6000 ! I also having data classes with lazy fields are little bit of overkill. As lazy values evaluated later they should never be saved and so better to keep them outside of the data model classes.
